rfcs/RFC-Upload-From-Dropbox.md

```diff
@@ -0,0 +1,417 @@
+[Marcus Kinsella](mailto:mkinsella@chanzuckerberg.com), [Eduardo Lopez](mailto:elopez@chanzuckerberg.com), [Brian Raymor](mailto:braymor@chanzuckerberg.com)
+[Ryan King](mailto:ryan.king@chanzuckerberg.com), [Brian McCandless](bmccandless@chanzuckerberg.com)
+
+## tl;dr
+
+A user can upload files, up to 30GB in size, to Data Portal from a DropBox link.
+
+## Glossary
+
+- Shared Link - A link from DropBox that gives the bearer of that link access to private data.
+- DP(Data Portal) - The cellxgene Data Portal.
+- Small File - a file < 100 MB in size.
+- Large File - a file > 100 MB in size.
+
+## Problem Statement | Background
+
+One of the primary goals of the DP is to allow users to upload their own files. These files consist of datasets.
+These datasets can vary in size from a few MB to several GB. Users can provide a shared link from their DropBox that
+is used to upload directly to S3.
+
+This specification does not cover the validation of the data or the process for a user to complete a submission.
+
+## Product Requirements
+
+1. A user can upload a file to DP using a shared link from their Dropbox.
+1. A user can initiate an upload from the DP Browser App.
+1. A user can abort an upload from the DP Browser App.
+1. A user can check the status and progress of their upload from the DP Browser App.
+1. A user can restart a failed upload from the DP Browser App.
+1. The user must be authenticated before uploading.
+1. A user who has not yet agreed to the Portal policies must not be able to upload their file.
+1. Supports the upload of AnnData, Loom, or Seurat v3 RDS. See [Reference](https://github.com/chanzuckerberg/corpora-data-portal/blob/main/backend/schema/corpora_schema.md#implementations).
+
+### Nonfunctional requirements
+
+1. Cloud to Cloud uploads will retry if they fail. Retry at least 5 times with an exponential back off.
+1. A user can upload a single file up to 30GB in size.
+1. Shared links referring to more than one file will be rejected.
+1. The status of an upload must be tracked and queryable.
+
+## Detailed Design | Architecture | Implementation
+
+### Cloud to Cloud Upload Flow
+
+Cloud to cloud upload requires the user to provide a shareable link for a file on their DropBox from which DP can upload
+the file to the S3 bucket. Now to walk through the cloud to cloud flow, see figure below.
+
+![Cloud to Cloud](https://app.lucidchart.com/publicSegments/view/e1e72143-c108-4c7f-bf6c-053d98256c1f/image.png)
+**Figure:** Architecture of a user uploading from their cloud to our cloud.
+
+#### 1. Get Share Link
+
+The user must generate a sharable link from their DropBox. Sharable links should only point to a single file, not a
+directory of files.
+
+#### 2. Share Link
+
+Using the browser App, make a _POST {submission_id}/upload/link_request to upload the file
+by providing the sharable link in the request body. The response will contain the dataset_id, that can be used to
```

Comment:
How does the dataset table in the main DP database find out about the new _dataset_id_?

Reply:
During the post request the dataset_id is added to the main database and the upload table. We need to add it to the main table to make sure the dataset_id is unique and also reserved for this upload.